lib/utils.py

```diff
@@ -12,6 +12,22 @@
 
 CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
 
+def is_outside_symlink(base_dir, symlink_path):
+    try:
+        abs_target = os.path.realpath(symlink_path)
+        return not abs_target.startswith(os.path.realpath(base_dir)), abs_target
+    except OSError:
+        return False, None  # Not a symlink
+
+def find_outside_symlinks(base_dir):
+    for root, dirs, files in os.walk(base_dir):
```

Comment:
I think I might have found another attack. Not quite sure how to exploit it though right now. os.walk does not follow symlinks: https://docs.python.org/3.9/library/os.html?highlight=os%20walk
```
By default, walk() will not walk down into symbolic links that resolve to directories. Set followlinks to True to visit directories pointed to by symlinks, on systems that support them.
```
so maybe you could point to a dir that you can then somehow have a symlink in that will then not be checked. I am not going to spend too much time on this but maybe if you check in a dir where the permissions don't allow anything and then because you are root in the container you can access this dir. But I haven't tried it

Reply:
good point to think about. But I think from a security point it does not matter. What we want to know is if the symlink goes out of the directory. As soon as walk finds one and it goes that will be alerted.

The user can link internally of the repository as much as they want. Since to be an effective escape they must link to the outside at some point. So even if one links to another symlink internally as long as THAT target does not link out it does not matter if we follow symlinks or not.

Actually it would be the other way round and you could trigger an endless recursion. So I would argue os.walk is fine

Do you agree?